yellow-paper/docs/public-vm/avm.md

```diff
@@ -162,3 +162,145 @@ INITIAL_MESSAGE_CALL_RESULTS = MessageCallResults {
 ```
 
 > Note: unlike memory in the Ethereum Virtual Machine, uninitialized memory in the AVM is not readable! A memory cell must be written (and therefore [type-tagged](./state-model#types-and-tagged-memory)) before it can be read.
+
+## Execution
+With an initialized context (and therefore an initial program counter of 0), the AVM can execute a message call starting with the very first instruction in its bytecode.
+
+### Program Counter and Control Flow
+The program counter (machine state's `pc`) determines which instruction to execute (`instr = environment.bytecode[pc]`). Each instruction's state transition function updates the program counter in some way, which allows the VM to progress to the next instruction at each step.
+
+Most instructions simply increment the program counter by 1. This allows VM execution to flow naturally from instruction to instruction. Some instructions ([`JUMP`](./InstructionSet#isa-section-jump), [`JUMPI`](./InstructionSet#isa-section-jumpi), `INTERNALCALL`, `INTERNALRETURN`) modify the program counter based on inputs.
+
+> Note: program counter will never be assigned to a value from memory. Jump destinations can only be constants from the contract bytecode.
```

Comment:
This is not true on a technicality as the call stack is implemented in memory

Reply:
Ah okay. Is that memory reserved only for the callstack? Or could it be modified by other instructions?

Reply:
The idea is that it that there will be a special area, no other instructions should be able to write into that region

Reply:
Removed note

Reply:
:thinking: Is the "call stack is implemented in memory" true for the high-level VM spec? Or is that a circuit implementation detail? In other words, should the internal callstack just become another field in `MachineState` in this spec? And it is a circuit implementation detail that it shares the same memory trace as main memory?

Reply:
For now, I put the following paragraph in there, but still wondering ^.

 "`JUMP`, `JUMPI`, and `INTERNALCALL` assign a new value to program counter from a constant present in the bytecode. These instructions never assign a value from memory to program counter. Before jumping, the `INTERNALCALL` instruction pushes the current program counter to an internal call-stack that is maintained in a reserved region of memory. `INTERNALRETURN` pops a destination from that internal call-stack and jumps there. Thus, jump destinations, can be either constants from the contract bytecode, or destinations popped from the internal call-stack.".